### ISSUE TITLE
Feature Request: Clearly document what each grpc thread is doing and how to configure the number of threads used by the library.

### ISSUE THREAD

**Issue by eladm-ultrawis** [grpc/grpc#32770]
I am using grpc async api, and I have my own infrastructure that handles grpc completion queue(s).

grpc used to create additinal 3 threads in my application when initializing the server. I could live with that.

I recently upgraded to a more recent version of GRPC 1.51.1 via vcpkg. now the same code initializes gazillion redundant threads that sit on a condition variable doing pretty much nothing other than wasting resources.

There should be a clear documentation on how to configure the number of threads used by the library, with a clear explanation of the purpose of each thread. Hopefully there should also be a way to use my own threads / completion queues without the library doing parallel things behind my back.

(I also know that there is the EventEngine API, but the documentation is non existing, or I could not find it).

**Comment by maximli** [comment]
I also have the same problem, is there a solution.

**Comment by drfloob** [comment]
> initializes gazillion redundant threads that sit on a condition variable doing pretty much nothing

Can you be more precise about "a gazillion" threads? There number of threads should be capped around `min(# of cores, 32)`. If it's drastically more than 32, I'd be curious to investigate if you can provide a min repro.

> Hopefully there should also be a way to use my own threads / completion queues without the library doing parallel things behind my back.

There is such an API, called the [EventEngine](https://github.com/grpc/grpc/blob/ffafac3ce8ea53cbceffb53063070be813edc31f/include/grpc/event_engine/event_engine.h). If you want full control over threading behavior in gRPC internals, implementing a custom EventEngine is your best path forward. It's no small endeavor for most, but the API docs are thorough and the reference implementations should be a good guide. We are still in the process of integrating this API into the core library, better documentation will arrive this year.

**Comment by eladm-ultrawis** [comment]
@drfloob by "gazillion" I mean something that is relative to the number of cores on my machine. which is quite a lot considering the small amount of computational work that grpc does in my case.

There seems to be an implementation change in recent versions that significantly increases the number of threads used by grpc. 

Generally speaking, I think it is very undesired for a library to open threads according to the number of cores. Unless we assume threads are free - which they are not.  Or we assume that users who uses grpc don't have more important things to do with their CPU cores.

I think there should be a way to control the number of threads that grpc spawns.

**Comment by drfloob** [comment]
> There seems to be an implementation change in recent versions that significantly increases the number of threads used by grpc.

Yes, as I mentioned in my previous comment, we're changing the underlying I/O implementations. If you need deeper control of gRPC's internal threading behavior, writing a custom EventEngine is one path forward for you, your application will have complete control of threading behavior. We will likely change gRPC's provided EventEngine implementations to optimize performance and resource utilization, so the default threading behavior is expected to change over time.

> Or we assume that users who uses grpc don't have more important things to do with their CPU cores.

If your application is not doing much with gRPC, having ~30 idle threads waiting on a condvar to be signaled should not be significantly tying up your CPUs. It's not ideal, but it should not be noticeable in most scenarios. Unless you are hitting a maximum number of system thread constraint, I'd be curious to see some results of a smaller thread count making a discernible difference on your system. You can manually make a change here to test it https://github.com/grpc/grpc/blob/a2c89d0b24ccdf86aa4b96572116d6eb3eb001bd/src/core/lib/event_engine/thread_pool.h#L133

**Comment by eladm-ultrawis** [comment]
@drfloob one can always argue that having idle threads has little overhead on overall performance. But in my view this is something that should be at least configurable. Just like opencv has cv::setNumThreads, or ONNXRuntime allows a user to limit the number of threads used by their thread pool. If I need my CPU cores for more important things, I would like to be able to limit how 3rd party libraries are using them. limiting the number of threads is the easiest way to do that.

**Comment by Tradias** [comment]
One thing I have noticed is that gRPC will create CPU-core-count many threads when using one of the services that are implemented in terms of the Callback API, for example `grpc::EnableDefaultHealthCheckService` or the ORCA service. That is, because the creation of the Callback completion queues comes with the creation of CPU-core-count many threads in the background. In asio-grpc I have reimplemented [the former](https://github.com/Tradias/asio-grpc/blob/master/src/agrpc/health_check_service.hpp) using solely CompletionQueue API to work around that behavior. Now, I don't think it is ideal to have to do so just to retain control over the number of threads that gRPC creates. I hope that the EventEngine provides a solution here. Additionally, it would be nice if the EventEngine could also make integration with asio/std::execution easier, but in its current state I don't see how it would achieve that.

**Comment by eladm-ultrawis** [comment]
@drfloob would it be possible to allow the users to provide a thread limit via some public grpc API? there is no sense in having so many threads do nothing.

![image](https://github.com/user-attachments/assets/d8c0913b-a42d-4494-bb99-51d633869be8)

**Comment by drfloob** [comment]
Closing this  as a duplicate request, see https://github.com/grpc/grpc/issues/28642

The default set of threads has changed, and will continue to change, since gRPC evolves to optimize for best performance. For those that need control over threading behavior, you can implement an [EventEngine](https://github.com/grpc/grpc/blob/ff9cc060907e88640a6fcfe78d0dfeb2f3d89ca1/include/grpc/event_engine/event_engine.h) with your desired concurrency mechanisms.
